File: python/src/openharness/result.py
"""Run results and the streaming handle."""

from __future__ import annotations

import asyncio
from collections.abc import AsyncIterator
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any

from .types import Message, StreamEvent, Usage

if TYPE_CHECKING:
    from .agent import Agent
    from .context import CancelToken
    from .tracing import Trace


@dataclass
class RunResult:
    """What a run produced.

    ``output`` is the parsed structured output when the agent has an
    ``output_type``, otherwise the final text. ``new_messages`` holds every
    message created during the run (the input included).
    """

    output: Any
    text: str
    new_messages: list[Message]
    last_agent: Agent
    usage: Usage = field(default_factory=Usage)
    turns: int = 0
    tool_calls: int = 0
    trace: Trace | None = None
    history: list[Message] = field(default_factory=list)  # messages loaded from the session

    def to_messages(self) -> list[Message]:
        """Full conversation (history + this run), ready to pass as the next input."""
        return [*self.history, *self.new_messages]

    def __str__(self) -> str:
        return self.text


class RunStream:
    """Handle returned by ``run_stream``. Iterate for events, then await ``result()``.

    stream = run_stream(agent, "Tell me a story")
    async for event in stream:
        if event.type == "text_delta":
            print(event.data["delta"], end="")
    result = await stream.result()
    """

    def __init__(self, gen: AsyncIterator[StreamEvent], cancel_token: CancelToken):
        self._gen = gen
        self._cancel = cancel_token
        self._result: RunResult | None = None
        self._error: BaseException | None = None
        self._done = asyncio.Event()

    def __aiter__(self) -> AsyncIterator[StreamEvent]:
        return self._iterate()

    async def _iterate(self) -> AsyncIterator[StreamEvent]:
        try:
            async for event in self._gen:
                if event.type == "run_end":
                    self._result = event.data["result"]
                yield event
        except BaseException as e:
            self._error = e
            raise
        finally:
            self._done.set()

    async def text(self) -> AsyncIterator[str]:
        """Only the text deltas."""
        async for event in self:
            if event.type == "text_delta":
                yield event.data["delta"]

    async def result(self) -> RunResult:
        """Wait for the run to finish (consuming any remaining events)."""
        if not self._done.is_set():
            async for _ in self:
                pass
        if self._error:
            raise self._error
        assert self._result is not None
        return self._result

    def cancel(self, reason: str = "cancelled") -> None:
        self._cancel.cancel(reason)

File: typescript/src/result.ts
/** Run results and the streaming handle. */

import type { Agent } from "./agent.js";
import type { CancelToken } from "./context.js";
import type { Trace } from "./tracing.js";
import type { Message, StreamEvent, Usage } from "./types.js";

export interface RunResult<TOutput = string> {
  /** Parsed structured output when the agent has an `outputType`, otherwise the final text. */
  output: TOutput;
  text: string;
  /** Every message created during the run (the input included). */
  newMessages: Message[];
  /** Messages loaded from the session before the run. */
  history: Message[];
  lastAgent: Agent<any>;
  usage: Usage;
  turns: number;
  toolCalls: number;
  trace?: Trace;
}

/** Full conversation (history + this run), ready to pass as the next input. */
export const toMessages = (r: RunResult<any>): Message[] => [...r.history, ...r.newMessages];

/**
 * Handle returned by `runStream`. Iterate for events, then await `result()`.
 *
 *   const stream = runStream(agent, "Tell me a story");
 *   for await (const ev of stream) if (ev.type === "text_delta") process.stdout.write(ev.data.delta);
 *   const result = await stream.result();
 */
export class RunStream<TOutput = string> implements AsyncIterable<StreamEvent> {
  private final?: RunResult<TOutput>;
  private error?: unknown;
  private finished = false;
  private iterator: AsyncIterator<StreamEvent>;

  constructor(
    gen: AsyncGenerator<StreamEvent>,
    private readonly token: CancelToken,
  ) {
    this.iterator = gen;
  }

  async *[Symbol.asyncIterator](): AsyncIterator<StreamEvent> {
    try {
      for (;;) {
        const { value, done } = await this.iterator.next();
        if (done) break;
        if (value.type === "run_end") this.final = value.data.result;
        yield value;
      }
    } catch (e) {
      this.error = e;
      throw e;
    } finally {
      this.finished = true;
    }
  }

  /** Only the text deltas. */
  async *text(): AsyncIterable<string> {
    for await (const ev of this) if (ev.type === "text_delta") yield ev.data.delta as string;
  }

  /** Wait for the run to finish (consuming any remaining events). */
  async result(): Promise<RunResult<TOutput>> {
    if (!this.finished) {
      for await (const _ of this) {
        /* drain */
      }
    }
    if (this.error) throw this.error;
    return this.final!;
  }

  cancel(reason = "cancelled"): void {
    this.token.cancel(reason);
  }
}
